File: mqtt/Tracking.py
""" import paho.mqtt.client as mqtt
from web3 import Web3
import time

# MQTT broker details
broker_address = "your_broker_address"
port = 1883
topic = "animal_tracking"

# Ethereum details
ethereum_provider = "http://your_ethereum_node_provider"
contract_address = "your_contract_address"
private_key = "your_private_key"

# Callback when the client connects to the broker
def on_connect(client, userdata, flags, rc):
    print(f"Connected with result code {rc}")
    # Subscribe to the topic when connected
    client.subscribe(topic)

# Callback when a message is received from the broker
def on_message(client, userdata, msg):
    data = msg.payload.decode()
    print(f"Received message: {data}")

    # Send the data to the Ethereum smart contract
    send_to_ethereum(data)

# Connect to Ethereum
w3 = Web3(Web3.HTTPProvider(ethereum_provider))
contract_address = w3.toChecksumAddress(contract_address)
contract_abi = [...]  # Insert your contract ABI here

contract = w3.eth.contract(address=contract_address, abi=contract_abi)

def send_to_ethereum(data):
    # Check if the Ethereum node is connected
    if w3.isConnected():
        # Unlock your Ethereum account
        w3.eth.default_account = w3.eth.account.from_key(private_key)

        # Send data to the smart contract
        transaction_hash = contract.functions.storeCaptureData(data).transact()
        transaction_receipt = w3.eth.waitForTransactionReceipt(transaction_hash)

        print(f"Transaction confirmed. Gas used: {transaction_receipt['gasUsed']}")
    else:
        print("Error: Could not connect to Ethereum node.")

# Initialize MQTT client
client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message

# Connect to the broker
client.connect(broker_address, port, 60)

try:
    while True:
        # Wait for 3 hours
        time.sleep(3 * 60 * 60)

        # The script will keep running until interrupted
except KeyboardInterrupt:
    print("Script terminated by user")
    client.disconnect()
    client.loop_stop() """





"""codigo abaixo é da pergunta: can you show me a python code
with mqtt protocol that receive animal tracking data from a IOT device?
The tracking data must be received each 3 hours"""




""" 
import paho.mqtt.client as mqtt
import time

# MQTT broker details
broker_address = "your_broker_address"
port = 1883
topic = "animal_tracking"

# Callback when the client connects to the broker
def on_connect(client, userdata, flags, rc):
    print(f"Connected with result code {rc}")
    # Subscribe to the topic when connected
    client.subscribe(topic)

# Callback when a message is received from the broker
def on_message(client, userdata, msg):
    print(f"Received message: {msg.payload.decode()}")

# Initialize MQTT client
client = mqtt.Client()
client.on_connect = on_connect
client.on_message = on_message

# Connect to the broker
client.connect(broker_address, port, 60)

# Loop to handle incoming messages and keep the script running
client.loop_start()

try:
    while True:
        # Wait for 3 hours
        time.sleep(3 * 60 * 60)
except KeyboardInterrupt:
    print("Script terminated by user")
    client.disconnect()
    client.loop_stop()
 """
 
 

File: mqtt/Tracking.js
const mqtt = require('mqtt');
const Web3 = require('web3');
const cron = require('node-cron');
const contractABI = require('./AnimalTrackingABI.json'); // Replace with your ABI file

const mqttBrokerUrl = 'mqtt://your-mqtt-broker-url';
const mqttTopic = 'animal-tracking-topic';

const web3 = new Web3('https://sepolia.infura.io/v3/cc5cf625859f4030bca941d5fee94d03'); // Replace with your Infura API key

const contractAddress = '0xYourContractAddress'; // Replace with your contract address
const contract = new web3.eth.Contract(contractABI, contractAddress);

const mqttClient = mqtt.connect(mqttBrokerUrl);

mqttClient.on('connect', () => {
    console.log('Connected to MQTT broker');
    scheduleDataCapture();
});

function scheduleDataCapture() {
    // Schedule the data capture job to run every 3 hours
    cron.schedule('0 */3 * * *', () => {
        console.log('Fetching and processing tracking data...');
        fetchDataAndProcess();
    });
}

function fetchDataAndProcess() {
    // Implement the logic to fetch tracking data from MQTT and process it
    mqttClient.subscribe(mqttTopic);
}

mqttClient.on('message', (topic, message) => {
    // Assuming the message is in JSON format
    const trackingData = JSON.parse(message);

    // Invoke the captureData function in the smart contract
    contract.methods.storeCaptureData(
        trackingData.animalId,
        trackingData.latitude,
        trackingData.longitude
    ).send({
        from: '0xYourSenderAddress', // Replace with your sender address
        gas: 200000 // Set an appropriate gas limit
    }).then((receipt) => {
        console.log('Transaction receipt:', receipt);
    }).catch((error) => {
        console.error('Error sending transaction:', error);
    });
});
